src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,629 @@
+    };
+    let acc = Access2 { ptr: PtrSelector::X, kind: AccessKind::Read };
+    let target = source.perform_access2(&acc, xy_rel).unwrap();
+    assert!(source.y.state.permission.is_reserved(None));
+    assert!(target.y.state.permission.is_frozen());
+    assert!(!source.distinguishable::<(), ()>(&target, xy_rel))
+}
+
+#[test]
+#[should_panic]
+/// This test in a generalization of `reserved_frozen_protected_distinguishable`
+/// that more exhaustively explores the feasability of inserting a spurious read.
+/// For a number of interleavings of different pieces of code involving
+/// two mutually foreign tags, we test if a spurious read through one
+/// changes the outcome (introduces UB in the target when there is none
+/// in the source).
+fn spurious_read() {
+    // The general code pattern being exhaustively tested is
+    //     [thread 1]             || [thread 2]
+    //     retag x (protect)      || retag y (protect/noprotect)
+    //                      arbitrary code (incl. optionally ret y)
+    //     <spurious read x> (*)  ||
+    //     Check if x and y are distinguishable between pre- and post-spurious read (**)
+    //
+    // (*): `x` must still be protected at the moment the spurious read is inserted
+    //      because spurious reads are impossible in general on unprotected tags
+    // (**): this includes a possible loss of protector for `x`, because we are now
+    //       considering in distinguishability an arbitrary continuation of the program
+    //       after the spurious read has already been done
+    //
+    // (we model the arbitrary racing code with two accesses on each side,
+    // with fixed interleaving, which produces all possibilities of interleavings
+    // up to repeated applications of idempotent transitions)
+
+    #[derive(Clone, Debug)]
+    struct Pattern {
+        xy_rel: RelPosXY,
+        x_retag_first: bool,
+        x_initial_state: LocationState,
+        y_initial_state: LocationState,
+        y_protected: bool,
+        y_return: bool,
+    }
+
+    impl Exhaustive for Pattern {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            let mut v = Vec::new();
+            for xy_rel in RelPosXY::exhaustive() {
+                for (x_initial_state, y_initial_state) in
+                    <(LocationState, LocationState)>::exhaustive()
+                {
+                    // We can only do spurious reads for initialized locations anyway.
+                    precondition!(x_initial_state.initialized);
+                    // Moreover we don't create `Active` or `Disabled` directly on reborrow.
+                    precondition!(!x_initial_state.permission.is_disabled());
+                    precondition!(!y_initial_state.permission.is_disabled());
+                    precondition!(!x_initial_state.permission.is_active());
+                    precondition!(!y_initial_state.permission.is_active());
```

Comment:
These are the states *before* the retags, right? Why can't those be active/disabled?

Reply:
Those are the states that `x` and `y` should have immediately after their (optional) retag.

Reply:
Oh, so they are not "initial" then, they are after-retag... that should be better reflected in comments and code. :)

Reply:
Sure, I can add a comment to make it explicit, but after-retag *is* initial since `x` and `y` don't even exist before the retag...

Reply:
From all the comments I was expecting "initial state" to refer to the state at the beginning of the example code. After all there's `before_arbitrary_code` to go from the initial state to the beginning of the arbitrary code.